tests/swift/benchmarks/Sources/benchmarks/main.swift

```diff
@@ -72,6 +72,24 @@ benchmark("structs") {
   fb.finish(offset: root)
 }
 
+// function to be used to avoid possible compiler optimization
+// which will just eliminate a block of code result of which not used
+@inline(never)
+func blackHole<T>(_: T) {}
+
+benchmark("load") {
+  let size = (1024 * 4)
+  var bytes: [UInt8] = []
+  for i in 0..<size-1 { bytes.append(UInt8(truncatingIfNeeded: i)) }
+  let bb = ByteBuffer(bytes: bytes)
+  var sum: UInt32 = 0
+  for i in 0..<size/MemoryLayout<UInt32>.size {
+    let pos = (i * MemoryLayout<UInt32>.size)
+    sum = sum &+ bb.read(def: UInt32.self, position: pos)
+  }
+  blackHole(sum)
+}
+
```

Comment:
Which read is this function hitting? the loadAligned or the unAligned?

Reply:
On the use-load-unaligned read use 'loadUnaligned'
On the master there is 'load' now.
I tried to run the benchmark on both use-load-unaligned and master branches and got same result for both.

Reply:
I checked out the repo and added two baseline benchmarks:

```
benchmark("load aligned baseline") {
    let size = (1024 * 4)
    let memory = UnsafeMutableRawPointer.allocate(byteCount: size, alignment: 1)
    for i in 0..<size-1 { memory.advanced(by: i).storeBytes(of: UInt8(truncatingIfNeeded: i), as: UInt8.self) }
    var sum: UInt32 = 0
    for i in 0..<size/MemoryLayout<UInt32>.size {
        let pos = (i * MemoryLayout<UInt32>.size)
        sum = sum &+  memory.advanced(by: pos).load(as: UInt32.self)
    }
    blackHole(sum)
}

benchmark("load unaligned baseline") {
    let size = (1024 * 4)
    let memory = UnsafeMutableRawPointer.allocate(byteCount: size, alignment: 1)
    for i in 0..<size-1 { memory.advanced(by: i).storeBytes(of: UInt8(truncatingIfNeeded: i), as: UInt8.self) }

    var sum: UInt32 = 0
    for i in 0..<size/MemoryLayout<UInt32>.size {
        let pos = (i * MemoryLayout<UInt32>.size)
        sum = sum &+  memory.advanced(by: pos).loadUnaligned(as: UInt32.self)
    }
    blackHole(sum)
}
```

and run them on my machine: Mac mini (M1, 2020)

Here are the results:

```
load                       18583.000 ns ±   2.50 %       75005
load aligned baseline        167.000 ns ± 353.87 %     1000000
load unaligned baseline      375.000 ns ± 2087.58 %    1000000
```

Seems like `loadUnaligned` is about 2x slower then `load`, but load through `ByteBuffer` is >100x slower, shadowing any kind of penalty `load` vs `loadUnaligned` might have.

Please review my findings as it might be that I missed something.

Reply:
I changed the benchmark a bit more to remove the creation of the data from the benchmark:

```
var globalSum: UInt32 = 0

let size = (1024 * 4)
let bb = {
    var bytes: [UInt8] = []
    for i in 0..<size-1 { bytes.append(UInt8(truncatingIfNeeded: i)) }
    return ByteBuffer(bytes: bytes)
}()


benchmark("load") {
  var sum: UInt32 = 0
  for i in 0..<size/MemoryLayout<UInt32>.size {
    let pos = (i * MemoryLayout<UInt32>.size)
    sum = sum &+ bb.read(def: UInt32.self, position: pos)
  }
  globalSum = sum
}

let bytes = {
    let memory = UnsafeMutableRawPointer.allocate(byteCount: size, alignment: 1)
    for i in 0..<size-1 { memory.advanced(by: i).storeBytes(of: UInt8(truncatingIfNeeded: i), as: UInt8.self) }
    return memory
}()



benchmark("load aligned baseline") {
    var sum: UInt32 = 0
    for i in 0..<size/MemoryLayout<UInt32>.size {
        let pos = (i * MemoryLayout<UInt32>.size)
        sum = sum &+ bytes.advanced(by: pos).load(as: UInt32.self)
    }
    globalSum = sum
}

benchmark("load unaligned baseline") {
    var sum: UInt32 = 0
    for i in 0..<size/MemoryLayout<UInt32>.size {
        let pos = (i * MemoryLayout<UInt32>.size)
        sum = sum &+  bytes.advanced(by: pos).loadUnaligned(as: UInt32.self)
    }
    globalSum = sum
}
```

And the results are:
```
load                       14958.000 ns ±   5.05 %      93339
load aligned baseline         83.000 ns ±  54.75 %    1000000
load unaligned baseline      334.000 ns ±  38.87 %    1000000
```

The difference between `load` and `loadUnaligned` is a bit prominent now at about 4x. And the difference to `ByteBuffer` load got even worse at 180x.

BTW I had to replace `blackHole` call with an assignment to a global variable, because it got inlined in `"load aligned baseline"` benchmark even though there is the `@inline(never)` attribute.

Reply:
So, yes, unfortunately unaligned load does not behave well...
And the reason for that is generic ByteBuffer.read<>
Benchmark source code is in the separate module from the FlatBuffer module, and seems the compiler can't handler properly. I tried to add 
```
  @inline(__always)
  public func readUInt32(position: Int) -> UInt32 {
    _storage.memory.advanced(by: position).loadUnaligned(as: UInt32.self)
  }
```
and benchmark shows numbers even better than with load.
My guess is the compiler emit extra code because it does not know a type that function will be used later with, and generate most safe code.

I played a bit trying to force a compiler generate good without big changes.
The one possible way is to force it to specialise ByteBuffer.read<> with all types natively supported by FlatBuffers.
I am going to update the PR, please let me know what you think about that.

Reply:
Yes I had also the suspecion that the huge penalty we see with `ByteBuffer.read<>` is not being able to perform the whole-module optimisation. I looked up your changes and with `@_specialize(exported: true, ...)` the benchmarks on my machine look as following:

```
load                        1292.000 ns ±  46.95 %    1000000
load aligned baseline         83.000 ns ± 735.11 %    1000000
load unaligned baseline      375.000 ns ± 137.81 %    1000000
```
It is much better from what we had before, but it still significant and therefore shadows the difference between `load` and `loadUnaligned`. That sad generally it is a win for the library, but IMHO we should consider probably on a different PR what can be done about the general `ByteBuffer` overhead. 


For example in [FlatBuffersSwift](https://github.com/mzaks/FlatBuffersSwift) / [FlatBuffersSwiftCodeGen](https://github.com/mzaks/FlatBuffersSwiftCodeGen) I allow users to use the Builder/Reader as a library or download the files directly into the project. This way the reading/writing infrastructure code becomes part of users code base and whole-module optimisation can do its magic. Not sure if this would be a proper solution here, hence the suggestion to create a separate issue and discussion for this topic.